### DISCUSSION TITLE
Records: EqualityContract is a Type

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#3787] [Q&A] [answered]
Why is the equality contract on records a type?  How will it be used?

I kind of expected it to to return a static singleton instance of a synthesized IComparer<T>.

**Comment by JustNrik** [comment]
It's used to compare the type of the current instance, pretty much like doing `typeof(ThisClass) == obj.GetType()` on `bool Equals(ThisClass obj)` but using that property instead.

**Comment by TonyValenti** [comment]
So would it essentially use an

Activator.CreateInstance(EqualityContract).Equals(x, y)

to do a compare?  What would the synthesized compare code look like?

**Comment by CyrusNajmabadi** [comment]
> So would it essentially use an

No. 

It will do:

```c#
=> this.EqualityContract == other.EqualityContract && ... do the member equality checks ...
```

**Comment by zvrba** [comment]
Why is `EqualityContract` even needed when `GetType()` exists?

**Comment by theunrepentantgeek** [comment, marked as answer]
Adding `EqualityContract` allows for precise control over whether subtypes are equal, or not.

See #3137 for more information

> The difficulty with implementing value-based equality across a type hierarchy lies in ensuring symmetry - that the two values agree on the equality being applied. In order to ensure that, types with value-based equality (or any custom equality) that can participate in a hierarchy of mutually comparable objects must essentially agree on who implements their equality. They can do that by declaring a Type-valued virtual property `EqualityContract` in the root class of the hierarchy, with every derived type that alters equality overriding that to return its own type.

**Comment by alrz** [comment]
>overriding that to return its own type.

Technically that's what GetType returns.

>>every derived type that alters equality 

Is there any case that a derived type *doesn't* alter equality?

**Reply by theunrepentantgeek** [reply]
> Is there any case that a derived type doesn't alter equality?

Most certainly.

For one example, when NHibernate returns an instance you've loaded from the database, you actually get a subclass of the expected type - so instead of a `Person` you get (effectively) a `PersonProxy` where all the properties have been overridden with additional functionality (for dirty tracking, lazy loading, etc). Yet a `PersonProxy` and a `Person` _can_, and **_should_**, compare as equal under certain circumstances.

Mads touches on the subject in the blog post [Welcome to C# 9.0](https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/#value-based-equality-and-inheritance).
